### PR TITLE
certificate support

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

File: node/internal.ts [hunk 1]
```diff
@@ -914,3 +917,64 @@ export function _normalizeSeparators(p: string): string {
...
+// return base64encoded<filePath>:base64encoded<encryptionKey>
+// downstream vsts-node-api will retrieve the secret later
+function _exposeTaskLibSecret(secret: string): string {
+    if (secret) {
+        let encryptKey = crypto.randomBytes(256);
+        let cipher = crypto.createCipher("aes-256-ctr", encryptKey);
+        let encryptedContent = cipher.update(secret, "utf8", "hex");
+        encryptedContent += cipher.final("hex");
+
+        let storageFile = path.join(_getVariable('Agent.TempDirectory'), uuidV4());
```
Comment (TingluoHuang) [inline node/internal.ts]: which folder i should fallback to? System.defaultworkingdir?

File: powershell/VstsTaskSdk/ServerOMFunctions.ps1 [hunk 2]
```diff
@@ -465,6 +480,39 @@ function Get-WebProxy {
...
+#>
+function Get-ClientCertificate {
+    [CmdletBinding()]
+    param()
+
+    Trace-EnteringInvocation -InvocationInfo $MyInvocation
+    try
+    {
+        # Min agent version that supports client certificate
+        Assert-Agent -Minimum '2.120.0'
```
Comment (TingluoHuang) [inline powershell/VstsTaskSdk/ServerOMFunctions.ps1]: bump this later. :)

File: node/internal.ts (hunk 1 above)
Comment (bryanmacfarlane) [inline node/internal.ts]: can we update the other .taskkey location to temp as well?
Fall back to OS temp.  But that means the file names can't have guids.

File: powershell/VstsTaskSdk/ServerOMFunctions.ps1 [hunk 3]
```diff
@@ -465,6 +480,39 @@ function Get-WebProxy {
...
+
+        [string]$clientCert = Get-TaskVariable -Name Agent.ClientCertArchive
+        [string]$clientCertPassword = Get-TaskVariable -Name Agent.ClientCertPassword
+        
+        if((Test-Path -LiteralPath $clientCert -PathType Leaf)) {
+            return New-Object -TypeName System.Security.Cryptography.X509Certificates.X509Certificate2 -ArgumentList @($clientCert, $clientCertPassword)
```
Comment (ericsciple) [inline powershell/VstsTaskSdk/ServerOMFunctions.ps1]: add dll first?

File: powershell/VstsTaskSdk/ServerOMFunctions.ps1 [hunk 4]
```diff
@@ -369,10 +369,25 @@ function Get-VssHttpClient {
...
+
+        $clientCert = Get-ClientCertificate
+        if($clientCert){
+            $null = Get-OMType -TypeName 'Microsoft.VisualStudio.Services.WebApi.VssClientCertificateManager' -OMKind 'WebApi' -OMDirectory $OMDirectory -Require
+            $null = [Microsoft.VisualStudio.Services.WebApi.VssClientCertificateManager]::Instance.ClientCertificates.Add($clientCert)
+
+            $null = Get-OMType -TypeName 'Microsoft.VisualStudio.Services.WebApi.VssClientHttpRequestSettings' -OMKind 'WebApi' -OMDirectory $OMDirectory -Require
+            [Microsoft.VisualStudio.Services.WebApi.VssClientHttpRequestSettings]::Default.ClientCertificateManager = [Microsoft.VisualStudio.Services.WebApi.VssClientCertificateManager]::Instance
+
+            $Settings = [Microsoft.VisualStudio.Services.WebApi.VssClientHttpRequestSettings]::Default.Clone()
```
Comment (ericsciple) [inline powershell/VstsTaskSdk/ServerOMFunctions.ps1]: other thing

File: powershell/VstsTaskSdk/ServerOMFunctions.ps1 (hunk 3 above)
Comment (TingluoHuang) [inline powershell/VstsTaskSdk/ServerOMFunctions.ps1]: the type is in System.dll, i assume that always there. :D

File: powershell/VstsTaskSdk/ServerOMFunctions.ps1 (hunk 4 above)
Comment (TingluoHuang) [inline powershell/VstsTaskSdk/ServerOMFunctions.ps1]: fixed

**Pull request merged**